Test cases improvement for MKLDNN on Gluon

juliusshufan opened the pull request:
## Description ##
This PR is a "follow-up" of previously merged #10764 .
In this PR, we are covering:

1. Refine the cases on nn.Conv2D and change the input shape to hit the MKLDNN code path;
2. Adding more test cases cover other gluon layers, like BN, Dense/FC, Pooling, Deconv etc. from the "MKLDNN-specialty" perspective;
3. Data coverage cases for some gluon layers, such as Conv2D, BN, Concat etc.

## Checklist ##
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
All the changes is reflected by tests/python/mkl/test_mkldnn.py

## Comments ##
1. For the correctness check on gluon computation, it follows the design used by tests/python/unittest/test_gluon.py, and therefore, the helper functions defined in tests/python/unitest/common.py is also used.

marcoabreu commented:
Hi, thanks for adding these tests! Could you elaborate why we need backend specific tests in a front-end language? Please excuse me if I'm making a misassumption here, but the implementation should be transparent and always act the same.

This means we should not need any MKL specific tests for operators, considering their behaviour should be identical. From my point of view, there should rather be general tests for all operators (in the same style you just wrote them), and they should just succeed if we switch the backend to MKL. 

I'm afraid that we will run into inconsistencies if we start writing custom tests for each backend. Ideally, all backends should produce the exact same output and execute the same behaviour. This would then be verified with general operator tests and thus make custom backend tests obsolete

@szha @piiswrong @zheng-da am I right with my assumption?

juliusshufan commented:
@marcoabreu Thanks for your comments and asking, and I fully understand your concerns on the consistence. Let me try to explain:
The main purpose for these newly add cases is to cover the "specialty" of the from the perspective of the integration of mkldnn with MXNET as a back-end, and data coverage improvement, that includes:
**1.** From input shape perspective, as the "memory layout" of mkldnn for different mkldnn primitives, selective input shape is necessary for hitting the mkldnn path, these can be reflected by test cases "test_mkldnn_conv2d", "test_mkldnn_batchnorm", in these cases, the channel/filter number is all multipliers of 16, and these can effectively cover the mkldnn code path; For the test_mkldnn_batchnorm case, because of the taking Conv2D output computed by the mkldnn path as the input of the BN layer can hitting the mkldnn code path, I therefore use the Conv2D output as the input of BN layer;   
**2. **Because (some of) the computations via mkldnn involves memory layout record, I add a series of test cases following the patterns involving a ndarray call involving potential memory operation (such as slice, reshape) and a real computation(such as conv, bn, pooling, dence/fc etc.) , these cases targeting the correctness check of the memory layout is properly handled before/after a computation via mkldnn.

As the test cases described in 1&2 not only focusing on the computation correctness, is it possible to execute these cases only for MKLDNN-enabled build by the Jeckins script? 
@marcoabreu @zheng-da @szha @piiswrong @pengzhao-intel @TaoLv May I have all your comments?

Thanks.

marcoabreu commented:
I see, thanks a lot for elaborating! Does this mean that MKLDNN is only being used if the input is in a certain shape? What happens in unusual shapes? The problem with this test is that you are not able to verify whether MKLDNN has actually been hit or if another implementation was used, right?

We might have to start looking into CPP tests for MKLDNN specific tasks, considering these backends are designed to be transparent and a lot of information is abstracted away due to the hourglass C-API. It might be easier to validate your assumptions in CPP. What do you think? I don't feel strongly about it, but this is something we should look into in future.

juliusshufan commented:
@marcoabreu Thanks for your prompt reply.

For the input shape, limited by my knowledge to mkldnn, I think it might be more accurate to call it as the  "preferred" shape, taking the conv2d as am example, the computation on 16X channel/filter number might be fully boosted/benefiting from mkldnn as the commonly used channel/filter number for neural network context is multiplier of 16 (such as 64, 128, 256, 512 etc.), for the behavior of mkldnn to unusual shapes, I think @zheng-da @pengzhao-intel @TaoLv are more expertised in this area, and can give a better explanation than me.

For CPP cases, I fully agreed with you it is helpful for the mkldnn specific tasks, especially for boundary/corner cases,  at this situation, the input might be invalidated by framework and the mkldnn can't be hit. Though, the python cases are relatively fast to construct and can focus on the specialty of integration of mkldnn, especially from the data coverage and the cases involving different layers combination involving both a computation via mkldnn and native CPU implementation path. I think these python cases can be further elaborated and executed according to the existing pre-ci scheme of MXNET.

Meanwhile, for CPP cases, actually I am planning the cases as well, and I am very willing to work with you guys on the test improvement at this area.

What do you think?

Thanks.

zheng-da commented:
@marcoabreu i agree with you that we need cpp tests for MKLDNN to explicitly cover different input and output NDArrays. I'm writing such unit tests. These tests will cover more different cases and many of them are not covered by the python unit tests. However, writing such cpp unit tests is much more difficult and probably can only be used for testing a single operator. We'll need more people to write unit tests.
The tests on in this PR. https://github.com/apache/incubator-mxnet/pull/10651
I'll probably split the PR and submit another PR for adding cpp unit tests.

That's why the python tests can be useful. It can easily cover many different operator combinations, which may cause problems in MKLDNN and was never tested in our unit tests.

juliusshufan commented on tests/python/mkl/test_mkldnn.py:
```diff
@@ -95,122 +110,1147 @@ def __getitem__(self, key):
+    for i in range(len(chn_list)):
+        for j in range(len(kernel_list)):
+            net = Net(chn_list[i], kernel_list[j])
+            check_layer_forward(net, x)
```
@zheng-da this is a kind of data coverage test cases, specifically covering the 16X channel and different spatial size from 1 to 299(these sizes normally used with cnn networks on imagenet), the existing conv2d cases not covering the 16x channel and diversified spatial size. Thanks.

juliusshufan commented on tests/python/mkl/test_mkldnn.py:
```diff
@@ -29,6 +29,21 @@
+    with mx.autograd.record():
+        out2 = net(x_hybrid)
+        print out2
+    out2.backward()
```
@zheng-da , my understanding is the comparison already between with hybridize and w.o hybridize.
"out1" is based on the network not hybridzied, and with x.copy, I set x_hybrid as the hybridized network at line 40 and compute as out2. Is it expected?

juliusshufan commented on tests/python/mkl/test_mkldnn.py:
```diff
@@ -95,122 +110,1147 @@ def __getitem__(self, key):
+    check_layer_forward(net, x)
+
+
+def test_reshape_gmaxpooling():
```
I'll modify the case accordingly.

juliusshufan commented on tests/python/mkl/test_mkldnn.py:
```diff
@@ -95,122 +110,1147 @@ def __getitem__(self, key):
+    z = mx.sym.add_n(z, y)
+    exe = z.simple_bind(ctx=mx.cpu(), x=x_shape, y=y_shape)
+    out = exe.forward(is_train=False, x=x_npy, y=y_npy)[0]
+    assert_almost_equal(out[0].asnumpy()[0, 0, 0], 1.0)
```
@zheng-da I guesss you are saying tc test_mkldnn_ndarray_slice, my understanding is the purpose of this case is similar to the one from mine, the difference is mine checks between with hybridize and w.o. hybridize. what do you think?

juliusshufan commented:
@zheng-da, thanks for your time on code review, I submitted a new version addressing your comments, especially the pooling related cases are packed to 6 cases to cover the existing 2d-pooling gluon API.

Could you please review my changes?

Thanks.

juliusshufan commented:
@marcoabreu @szha I checked the preci logs, and looks it is not relevant to my changes. if @zheng-da is fine with the modifications to your comments? @marcoabreu may I know if this PR can be accepted?

marcoabreu commented:
Thanks a lot for adding so many tests, it's very appreciated!!!

juliusshufan commented:
@marcoabreu Thanks for your review and kind words. These cases focusing on Gluon API, for your requests  "please move them to the general operator tests", shall I move these cases to the test_gluon.py or test_operator.py. Thanks.

marcoabreu commented:
Gluon fits even better, good idea!

zheng-da commented:
could you rebase your code? it seems something is messed up

juliusshufan commented:
@zheng-da clean-up done. Thanks @marcoabreu

zheng-da commented:
why does it fail in so many tests?

juliusshufan commented:
@zheng-da all the failed cases are the newly added one.
I can also reproduce the failures with my own ci environment. 
For dense API, there is a known issue here: https://github.com/apache/incubator-mxnet/issues/10807
For others cases which the pattern like reshape, slice connecting to batchnorm and conv/deconv, the failure are most likely caused by reorder.
So, may I have all your opinion regarding to the failed cases? @marcoabreu

zheng-da commented:
@juliusshufan https://github.com/apache/incubator-mxnet/issues/10807 does this case work on CPU without mkldnn?

zheng-da commented:
your input size is too large for Dense. Very few systems can actually allocate a single piece of memory of 85GB. There is no reason to test the error in https://github.com/apache/incubator-mxnet/issues/10807. No system can handle it.

marcoabreu commented:
I think it's a good these tests failed - it means that we actually have some issues and are being inconsistent. That's a great point to start from.

I'd appreciate it if you could make the necessary changes in order to make all parts consistent.

zheng-da commented:
@marcoabreu i'm not sure what you mean. do you mean we should support NDArrays of 85GB? i don't think we can use malloc to allocate a single piece of memory of this size?

marcoabreu commented:
Aaaah no no, definitely not. It was just a general statement :)

Da Zheng <notifications@github.com> schrieb am Sa., 26. Mai 2018, 02:14:

> @marcoabreu <https://github.com/marcoabreu> i'm not sure what you mean.
> do you mean we should support NDArrays of 85GB? i don't think we can use
> malloc to allocate a single piece of memory of this size?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/incubator-mxnet/pull/10921#issuecomment-392218718>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ARxB68F9X4zhbl9p2tYinw1NKiIH18nTks5t2J55gaJpZM4T8smV>
> .
>

zheng-da commented:
@marcoabreu I think we should remove the tests for generating huge weight matrices, as @juliusshufan is doing right now for Dense.
It's clear that the code will run into out-of-memory errors, regardless of the backend (native MXNet, MKLDNN or CuDNN).

marcoabreu commented:
I think they were originally introduced to test int32 overflow, but I
totally agree that these are suboptimal and should be rather be mocked in
the C++ tests opposed to actually allocating so much memory.

Da Zheng <notifications@github.com> schrieb am Mo., 28. Mai 2018, 04:35:

> @marcoabreu <https://github.com/marcoabreu> I think we should remove the
> tests for generating huge weight matrices, as @juliusshufan
> <https://github.com/juliusshufan> is doing right now for Dense.
> It's clear that the code will run into out-of-memory errors, regardless of
> the backend (native MXNet, MKLDNN or CuDNN).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/incubator-mxnet/pull/10921#issuecomment-392403691>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ARxB65K-DpBIA66UA8wL2SxlG9xIEjWWks5t22JzgaJpZM4T8smV>
> .
>

juliusshufan commented:
@zheng-da @marcoabreu , I'll modify/remove the case causing oom, meanwhile, we are working on the bugfix for the remaining failed cases.

Thanks.

TaoLv commented:
Any other road blocks for this PR merge?
@marcoabreu Do you think we can merge these tests firstly, with those failed tests skipped? We can track them with JIRA tickets and re-open them after they are fixed. Actually, we have been working on these and hopefully will submit PR to fix them soon.

marcoabreu commented on tests/python/unittest/test_gluon.py:
```diff
@@ -1146,6 +1147,1046 @@ def test_summary():
+    mx.test_utils.assert_almost_equal(out1.asnumpy(), out2.asnumpy(), rtol=1e-5, atol=1e-6)
+
+@with_seed()
+def test_conv2d_mkldnn():
```
remove mkldnn in name?

marcoabreu commented on tests/python/unittest/test_gluon.py:
```diff
@@ -1146,6 +1147,1046 @@ def test_summary():
+
+
+@with_seed()
+def test_group_conv2d_mkldnn():
```
remove mkldnn in name?

marcoabreu commented on tests/python/unittest/test_gluon.py:
```diff
@@ -1146,6 +1147,1046 @@ def test_summary():
+@with_seed()
+def test_group_conv2d_mkldnn():
+    if mx.Context.default_ctx != mx.cpu(0):
+        raise SkipTest("This test case is only for CPU to avoid OOM on GPU!")
```
Please create a ticket and put it into the message

marcoabreu commented on tests/python/unittest/test_gluon.py:
```diff
@@ -1146,6 +1147,1046 @@ def test_summary():
+
+
+@with_seed()
+def test_deconv2d_mkldnn():
```
remove mkldnn in name?

marcoabreu commented on tests/python/unittest/test_gluon.py:
```diff
@@ -1146,6 +1147,1046 @@ def test_summary():
+    check_layer_forward_withinput(net, x)
+
+
+def test_reshape_conv_slice_conv():
```
add @with_seed

marcoabreu commented:
@TaoLv which tests exactly would you like to disable? Please describe for each test why it's failing.

Do you know by when you'd have a working version?

marcoabreu commented on tests/python/unittest/test_gluon.py:
```diff
@@ -1146,6 +1147,1046 @@ def test_summary():
+@with_seed()
+def test_reshape_conv_slice_conv():
+    """
+    This test will test gluon Conv2d computation on mkldnn with ndarray reshape and slice
```
MKLDNN mentioned

juliusshufan commented:
@marcoabreu Appreciate for your code review, on my own preci env, with lastest mxnet master code, there are in total 11 failed test cases (in test_gluon.py), they are:
test_deconv2d_16c
test_batchnorm_16c
test_reshape_conv_reshape_conv
test_slice_conv_reshape_conv
test_reshape_batchnorm
test_reshape_deconv
test_slice_deconv
test_reshape_deconv_reshape_deconv
test_slice_deconv_slice_deconv
test_reshape_deconv_slice_deconv
test_slice_deconv_reshape_deconv

From back-trace perspective, the failures can be ascribed to two kind of issues, I think I can raise issues in either github/JIRA to track the bugfixes with this PR(which adding the test cases) merged. What do you think? Thanks for your time.

@TaoLv

TaoLv commented:
Thanks @juliusshufan .
@marcoabreu These test cases are mainly for verifying the interchange between mxnet default memory layout and mkldnn layout. For those conv releated failed cases, we already have a fix in local, but still need full validation before submitting to master branch. For deconv releated cases, we will have a fix in this week.

marcoabreu commented:
Hmm, I'm a bit on the fence here. On one hand you're only introducing new tests, so it's not that bad. On the other hand, we'd like to avoid introducing disabling tests.

Just for my understanding. Do you want to get this PR merged since other PRs depend on it or what exactly is the background? In general, I'd rather prefer if we get tests to work before we commit them. If you got a good reason and can assure that you'll get the fixes in this week, I won't stand in the way.

TaoLv commented:
Thanks, @marcoabreu . I totally understand your concern on that. There is no other PR pending on this. I just want those good tests can go into ci for full validation and protect the existing code when another PR is submitted by other developers. So now we have two choices. One is keeping this PR open until we fix all issues and merge them to master branch. The other one is merging those passed cases firstly and leave those failed cases to the following bug fix PRs.
Which one do you prefer? @marcoabreu @juliusshufan

marcoabreu commented:
I'm on my phone right now. How many tests would be active after the merge?

Tao Lv <notifications@github.com> schrieb am Di., 5. Juni 2018, 12:11:

> Thanks, @marcoabreu <https://github.com/marcoabreu> . I totally
> understand your concern on that. There is no other PR pending on this. I
> just want those good tests can go into ci for full validation and protect
> the existing code when another PR is submitted by other developers. So now
> we have two choices. One is keeping this PR open until we fix all issues
> and merge them to master branch. The other one is merging those passed
> cases firstly and leave those failed cases to the following bug fix PRs.
> Which one do you prefer? @marcoabreu <https://github.com/marcoabreu>
> @juliusshufan <https://github.com/juliusshufan>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/incubator-mxnet/pull/10921#issuecomment-394655979>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ARxB6yZbYZzFyDUAv7pR3wN48ifanAlUks5t5llPgaJpZM4T8smV>
> .
>

TaoLv commented:
About 30 test cases would be active there.

marcoabreu commented:
Alright, lets create GitHub issues for the disabled tests and merge. Please put the link into the message that's being returned if a test is skipped so we can keep track of it.

TaoLv commented:
Thanks.
@juliusshufan Could you help to create a GitHub issue and refine this PR according to @marcoabreu's suggestion?

juliusshufan commented:
@TaoLv Sure and done.. @marcoabreu

juliusshufan commented:
@marcoabreu Hi Marc, may I know any action I need to take for the PR merged?

Thanks.

marcoabreu commented:
We're only waiting for CI :)

juliusshufan commented:
@marcoabreu Thanks for your prompt reply, seems the failures not caused by my commit. so for CI you are saying everything should pass, right ?

marcoabreu commented:
Right, sorry for the inconvenience.

juliusshufan commented:
@marcoabreu Hi Marc, I retriggerd the CI today, but still many failed cases. I looked into the failed cases, and I found the some of the failed cases are introduced by my PR, but some are not. For the failed cases introduced by my PR, they all failed on GPU test, taking the example of ci "ut-python-gpu", there are in total 109 failed cases, some of the cases from my PR, but I reproduce the cases on CPU with MKLDNN enabled and same seed, and the failure can't be reproduced. Limited by my knowleged to GPU, may I have your suggestions and expertise on how to moving-on. I assume the cases pass on CPU should be no issue on GPU as well...

Thanks in advance for your time.

marcoabreu commented:
It's very interesting that basically all GPU tests failed with mem access. Strange.. I have retriggered the tests :/

juliusshufan commented:
@marcoabreu Thanks for your retriggering, seems the result is very similiar... too many gpu cases failed.

marcoabreu commented:
I'm tempted to say that your tests might be revealing a memory leak/corruption in MXNet. 

Could you please try to reproduce these results locally? https://cwiki.apache.org/confluence/display/MXNET/Reproducing+test+results

zheng-da commented:
can we trigger the PR again?

juliusshufan commented:
@zheng-da , the retrigger will not work for this case. I am still working on the investigation why these cases failed on GPU, as some of my cases are data coverage purposes, and might consume more memory, not for sure if these kind of cases applicable for GPU test.

zheng-da commented:
can you try rebase and see if it works?

juliusshufan commented:
@marcoabreu Sorry for no update on this PR for a long time. I did some investigations and reproduce the same GPU tests on a NVidia P40 machine, and I think the previous failured reported from this CI is caused out of memory (though some of failures are "illegal memory"). I therefore modify my cases to reduce the tensor size, by using a smaller input shape, channel number, batch size etc. And on a P40 GPU, all the GPU cases, including test_gluon and the cases under tests/python/gpu passed.
Now you can see, the latest CI results show much fewer failed cases than before, though it still failed due to "out of memory". I might keep on reducing the tensor size to avoid the out of memory on the MXNET preci environment, while with the cases increase, eventually the out of memory will happen.

As we dicussed before, I agreed that it is better that test cases not bind to a backend, while these cases targeting the MKLDNN related layout and memory reorder, per the current ciucumstance, may I suggest to executed only on CPU? These cases are helpful to track the regressions on MKLDNN integration. @zheng-da @TaoLv

May I have your comments?

Thanks.

marcoabreu commented:
Hi, thanks for the detailed explanation. While illegal memory accesses could be indicating a bug in our implementation (we'd have to run with cuda memcheck to verify), out of memory errors should be straight forward and it should be no problem to reduce the tensor sizes. I just want to make you aware of the fact that we got 8GB of GPU memory available and that we don't run GPU tests in parallel. In case you'd like to reduce the sizes, please consider whether the tensors are actually too big or if there could be other problems in our implementation.

I'm not 100% familiar with the details, so please bare with me if my questions don't make sense. In general, I understand that you have some special optimized features in mkldnn which are triggered by certain tensor sizes and memory layouts. My question here is how the user is able to influence that: are they able to select a certain layout or is this something we do implicitly in the backend? If they are able to select it, but it's not available in the current backend (because it's not supported), how does the behaviour continue? Is there a fallback mechanism? What we should avoid is to add features that are only usable in very certain environments. In general, we would like our users to define their networks without the backend in mind and trust us that we are doing our best to execute it most effectively - or at least fall back to a method which will still deliver correct results but may be slower (we've been thinking about an Auto Tuning mechanism for the future to select the best implementation). Please understand that we want to avoid having a vendor lock in our models - users should not be restricted to MKLDNN, cuda oder cudnn if they design their networks.

If everything you'd like to test is completely happening in the background, you're free to make mkldnn specific tests. But in that case, I'd assume that they'd be cpp tests, considering the assumption that this does not depend on the frontend and thus would be ineligible for a Python test.

I'm making a lot of assumptions here, so please don't hesitate to correct me if something does not match the reality.

juliusshufan commented:
@marcoabreu thanks for you prompt reply, I modify the cases and rerun, similiar issues are there.
The GPU I used has more than 16G memory, and this might be the reason why I can't reproduce the issues.
I noticed, if I did not use a smaller tensor on purpose, and:
1. If I run the cases one by one,  like nosetests test_gluon: tc, then there is no issue.
2. If I run nosetests:test_gluon, then there is no issue.

The issue only happens at nosetests:test_operator_gpu, as the test_gluon.py is imported, the nosetests:test_operator_gpu will run many test cases. My assumption is, as long as the (test) purpose can be fullfiled, writhing python cases are fast than CPP, and since via the python API, and this is the cases I am writing...

Limited to my knowledge on GPU, may I suggest to involve a folk have more GPU expertise, thanks for your help.

juliusshufan commented:
@marcoabreu @szha @eric-haibin-lin @piiswrong @zheng-da
I move the cases to tests/python/mkl from tests/python/unittests, and per the previous dicussion with Marc, I think it is better to classify these cases for mkldnn backend but not unified for GPU and CPU, the justification is, the main purpose of these cases are checking if the MKLDNN-enabled Op can cope with a sliced or reshaped Ndarray; as MKLDNN sometimes invovles the specific memory layout, and this is a speciality of MKLDNN but not CUDA/CUDNN. Though this is a backend specific feature, since python test cases are qualified to fullfil the testing purpose, I think it make sense to keep it for mkldnn backend.

May I have your comments on how to move forward? Via these cases, valid issue(s) have been identified, and it is helpful to track the regression for MKLDNN integration with MXNET. Thanks.
BTW: I checked the CI failure with this commit, seems they are not relevant to the cases.

Thanks for all your time.

marcoabreu commented:
While I understand the intention, I don't see why the tests should be
limited to MKLDNN. You want to test certain shapes that then trigger a
special behaviour in the MKLDNN backend and that's totally fine and the
best way to do it! But what I don't understand is why we can't run this
test on the default MXNet backend on CPU or GPU. What would happen? Would
it fail? If it wouldn't (and I hope that it won't), I don't see much reason
why this should be a MKLDNN specific test. We're basically increasing the
coverage of the test by moving it out of the MKLDNN specialised tests
without sacrificing anything. On the other hand, we're even making sure
that things, that users are supposed to do with the MKLDNN backend, also
works for other backends. The backend is and should be entirely invisible
to our users and thus we should always ensure (on a best effort base) that
our features work on all backends. We employ the same policy to CUDA: Every
test that can be run on CPU also has to pass on GPU. MKLDNN is, from my
point of view, just another type, thus making it CPU-MXNet, CPU-MKLDNN and
GPU-CUDA. Means, every backend has to pass the same set of tests - starting
with exceptions like this one leaves room for diverging functionality of
the different backends. In the end, our users will be facing a bad
calculation or an unexpected error and then leave with a bad user
experience - we want to avoid that at all costs.

As stated in
https://github.com/apache/incubator-mxnet/pull/10921#pullrequestreview-121193502,
please move them to gluon test or extend the existing test cases with these
special cases. Feel free to add a comment for justification (e.g. why you
chose a specific shape etc).

lupesko commented:
@juliusshufan bouncing this PR, this is a great contribution, would be awesome if you can follow up on feedback and look into the build failure.
Thanks for your contribution!

marcoabreu commented:
We use AWS g3.8xlarge instances

pengzhao-intel commented:
Yes, it's a group of very useful test cases. I am looking into the case with @juliusshufan to resolve the issue.

pengzhao-intel commented:
The GPU issue is tracked by https://github.com/apache/incubator-mxnet/issues/12453 which blocked the merge of these test cases.

juliusshufan commented:
@marcoabreu Hi Marc, with revising the cases, right now all the test cases pass the CI, the skipped ones are tracked by https://github.com/apache/incubator-mxnet/issues/11164.

juliusshufan commented:
@marcoabreu Thanks for reopening the https://github.com/apache/incubator-mxnet/issues/11164 for tracking the temporarily skipped test cases in this PR, may I have your comments on the current status of this PR, since all CI failures cleared now... Sorry for pushing you.

juliusshufan commented:
@lupesko The CI issues are finnally cleared, may I know who can keep on the reviewing and moving forward. Thanks.

juliusshufan commented:
@marcoabreu Hi Marc, sorry for pushing you and I know you are busy, while I am still looking forward your reply... Thanks.

juliusshufan commented:
@szha Hi Sheng, this PR is part of MKLDNN integration (https://issues.apache.org/jira/browse/MXNET-425), and since Marc has previously approved, and there is no CI issues reported, may I know if this PR is qualified to merge?

Thanks for your time.

stu1130 commented:
@mxnet-label-bot[pr-awaiting-merge]

vandanavk commented:
@szha for review/merge

juliusshufan commented:
@vandanavk thanks for the labelling. @szha, may I have your time on the review/merge? This PR has been opened for pretty a while... Thanks in advance for your time.
